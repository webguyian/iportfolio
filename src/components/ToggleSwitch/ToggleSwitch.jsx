import React, { useState } from 'react';
import { func, string } from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import Text from 'components/Text/Text';

const ToggleSwitch = props => {
  const { label, onUpdate } = props;
  const [checked, setChecked] = useState(false);
  const baseClass = 'ui-toggle-switch';
  const activeClass = checked ? `${baseClass}--active` : false;

  const handleChange = ({ target }) => {
    setChecked(target.checked);
  };

  return (
    <label className={classNames(baseClass, activeClass)}>
      <input
        className={`${baseClass}-checkbox`}
        type="checkbox"
        checked={checked}
        onChange={handleChange}
      />
      <Button className={`${baseClass}-slider`} onTransitionEnd={onUpdate} />
      <Text className={`${baseClass}-label`}>{label}</Text>
    </label>
  );
};

ToggleSwitch.propTypes = {
  label: string,
  onUpdate: func.isRequired
};

ToggleSwitch.defaultProps = {
  label: 'slide to unlock'
};

export default ToggleSwitch;
