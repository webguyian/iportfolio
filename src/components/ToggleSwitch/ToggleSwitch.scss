.ui-toggle-switch {
  display: block;
  position: relative;
  margin: 0 auto;
  border-radius: 25px;
  background-color: rgba($black, 0.5);
  text-align: center;
  cursor: grab;

  &-checkbox {
    @include accessibly-hidden;
  }

  &-slider {
    display: block;
    position: absolute;
    top: 8px;
    left: 8px;
    width: 30px;
    height: 30px;
    transition: 0.3s transform;
    border-radius: 50%;
    background-color: $white;
    content: "";

    .ui-toggle-switch--active & {
      transform: translateX(170px);
    }
  }

  &-label {
    display: inline-block;
    padding: 10px 40px;
    transform: translateX(10px);
    transition: color 0.6s, transform 0.3s;
    color: $gray;
    font-size: $font-size-l;
    font-weight: $font-weight-light;

    .ui-toggle-switch--active & {
      transform: translateX(-10px);
      color: $white;
    }
  }
}
