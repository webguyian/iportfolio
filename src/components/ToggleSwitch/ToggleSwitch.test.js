import React from 'react';
import { act, create } from 'react-test-renderer';

import ToggleSwitch from './ToggleSwitch';

describe('<ToggleSwitch />', () => {
  const props = {
    onUpdate: jest.fn()
  };

  it('renders correctly', () => {
    const component = create(<ToggleSwitch {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('handles change', () => {
    const component = create(<ToggleSwitch {...props} />);
    const input = component.root.findByType('input');

    act(() => {
      input.props.onChange({ target: { checked: true } });
    });

    expect(component).toMatchSnapshot();
  });
});
