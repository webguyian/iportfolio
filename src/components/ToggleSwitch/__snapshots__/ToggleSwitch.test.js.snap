// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleSwitch /> handles change 1`] = `
<label
  className="ui-toggle-switch ui-toggle-switch--active"
>
  <input
    checked={true}
    className="ui-toggle-switch-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <button
    className="ui-btn ui-toggle-switch-slider"
    onTransitionEnd={[MockFunction]}
    type="button"
  />
  <span
    className="ui-text ui-text--body ui-toggle-switch-label"
  >
    slide to unlock
  </span>
</label>
`;

exports[`<ToggleSwitch /> renders correctly 1`] = `
<label
  className="ui-toggle-switch"
>
  <input
    checked={false}
    className="ui-toggle-switch-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <button
    className="ui-btn ui-toggle-switch-slider"
    onTransitionEnd={[MockFunction]}
    type="button"
  />
  <span
    className="ui-text ui-text--body ui-toggle-switch-label"
  >
    slide to unlock
  </span>
</label>
`;
